WIP: #319 Seasonal storage/rep. days

mihlema opened the pull request:
Current implementation allows us to have representative days + seasonal storages. node_state variables, investment variables are considered to be always non-representative, i.e. always exist for the full optimization period, while all others are representative, i.e. mapped for non-representative periods to their corresponding representative period.
Mapping of the variables happens in `variable_common`. The filtering of `rep` and `non-rep` temporal_blocks happens in the variable_indices functions. 
Additional features:
- dynamic constraint: dynamic constraints are only generated for rep. days, but for their overlap to previous period, we also take into account the representative period of the previous period. See e.g. `unit_dynamic_indices`
- to ease the implementation of multiple temporal_blocks, a temporal_block group can be used to connect to node. Members function is called within the code
Known issues:
- as stated before, node_state balance is always assumed to be non-representative,i.e. created for the full year. At the moment it's not possible to combine non-seasonal and seasonal storages. Would we need this?
- the node_injection constraint is created for the enitre year. However, for e.g. nodes without storages, it could also simply be created for only rep. days. I haven't found a way to implement this is a nice way
- in the `node_injection` constraint, the representative demand of the current timeslice needs to be called. The current implementation is working, but not very pretty.
- currently the temporal_block weights are summed up. This is a bug and needs to be changed to prod(...) . Currently this is not supported by Spineopt and interface

manuelma commented on src/run_spineopt.jl:
```diff
@@ -324,6 +324,31 @@ function optimize_model!(m::Model; log_level=3, calculate_duals=false, mip_solve
     # but is not mentioned in the COLUMNS section. We are ignoring it.
     @timelog log_level 0 "Optimizing model $(m.ext[:instance])..." optimize!(m)
+    if termination_status(m) == MOI.INFEASIBLE
+        compute_conflict!(m)
```
This is nice. Would you consider putting this code in its own function, and call it from here?

Also, just to be sure this is not related to the current MR? If not, would you consider committing it to master directly?

mihlema replied:
Yes, sorry this one was supposed to come in a separate MR. I'll try to remove it from this one and merge it separately. But it's indeed really cool, already saved me a bunch of time.

manuelma commented on src/variables/variable_common.jl:
```diff
@@ -52,9 +52,33 @@ function add_variable!(
+"""
+function map_variable!(m::Model, var::Dict, var_indices::Function)
+    # TODO: should probably rename "ordering_periods" to
+    # "time_slice_mapping"
```
I think `representative_periods` is a nice name. "Ordering" doesn't sound correct to me, sounds like something else.

`representative_periods` or `representative_time_slices` sounds good. `time_slice_mapping` is too general and doesn't say anything about 'representative'.

manuelma commented on src/variables/variable_unit_flow.jl:
```diff
@@ -35,18 +35,21 @@ function unit_flow_indices(
         for
         (n, s, t) in
-        node_stochastic_time_indices(m; node=n, stochastic_scenario=stochastic_scenario, temporal_block=tb, t=t)
+        node_stochastic_time_indices(m; node=n, stochastic_scenario=stochastic_scenario, temporal_block=temporal_block, t=t)
```
Why this change? I think it's more consistent to name the iteration variable `tb`, as it was before...

manuelma commented on src/variables/variable_common.jl:
```diff
@@ -52,9 +52,33 @@ function add_variable!(
+        Keys = [k for k in keys(ind) if !(typeof(ind[k]) <: TimeSlice)]
+        Values = [ind[k] for k in Keys]
+        non_t_slice_ind = (; zip(Keys, Values)...) #this gets everything, but the timeslice index...
+        ind_rep = first(var_indices(m;non_t_slice_ind...,t=rep_time_slices(m)[to_time_slice(m,t=ind.t)]))
```
Why do we need to call `to_time_slice` here? isn't always the case that `ind.t` is already in the model? I don't see how else it would have made it to a variable index (maybe I'm missing something?)

Because if I'm not mistaken `to_time_slice(m,t=ind.t) == t.ind.t`

manuelma commented on src/objective/variable_om_costs.jl:
```diff
@@ -30,6 +30,7 @@ function variable_om_costs(m::Model, t1)
         expr_sum(
             +unit_flow[u, n, d, s, t] *
             duration(t) *
+            sum(weight(temporal_block=blk) for blk in t.blocks) *
```
I have a small suggestion: replace `t.blocks` by `blocks(t)`. It's more api-friendly than accessing fields directly.

mihlema replied:
Yes that looks better. Also, related to the original MR message:
> currently the temporal_block weights are summed up. This is a bug and needs to be changed to prod(...) . Currently this is not supported by Spineopt and interface
I have tried to use prod(..) or Iterators.prod(..) here, but the iteration method seems to fail here. Do you have an idea on that? I guess alternatively reduce(*,...) could work

manuelma commented on src/variables/variable_common.jl:
```diff
@@ -52,9 +52,33 @@ function add_variable!(
             ind => _variable(m, name, ind, lb, ub, bin, int)
             for ind in indices(m; t=vcat(history_time_slice(m), time_slice(m)))
         )
+    if use_representative_days(model=m.ext[:instance]) == :use_representative_days_w_seasonal_storage
```
Is the spine_opt_template.json updated at this point? I think we'd need to add the use_representative_days parameter together with the possible 'enumerated' values. Which others would we have apart from use_representative_days_w_seasonal_storage? is the `w_seasonal_storage` part really necessary? Maybe it could be just a true/false thing.

mihlema replied:
Yes, I think we could have a boolean parameter `use_representative_days_w_seasonal_storage`. I think the `w_seasonal_storage` doesn't look really nice, but I think it'd be important to indicate it here. (For "normal" representative days we don't need the whole mapping thing).
Alternatively, we can also trigger the mapping by the existence of the `ordering_periods` parameter, I think. Then we wouldn't need the `use_representative_days_w_seasonal_storage` parameter at all.

manuelma commented on src/data_structure/temporal_structure.jl:
```diff
@@ -344,6 +345,28 @@ function to_time_slice(m::Model; t::TimeSlice)
+        end
+    end
+    m.ext[:temporal_structure][:rep_day_mapping] = rep_dict
+    m.ext[:temporal_structure][:non_rep_ind] = keys(rep_dict)
```
I don't see `:non_rep_ind ` being used anywhere else?

manuelma commented on src/variables/variable_common.jl:
```diff
@@ -52,9 +52,33 @@ function add_variable!(
 end
 
+"""
+    map_variable!(v::Dict{VariableRef}, indices_all::Function)
```
maybe the name could be `replace_non_representative_periods!(...)` or `map_periods!(...)`? I don't know, `map_variable!` sounds a bit odd. Maybe `remap_variable!` ???

manuelma commented on src/data_structure/temporal_structure.jl:
```diff
@@ -283,6 +283,7 @@ function generate_temporal_structure!(m::Model)
     _generate_current_window!(m::Model)
     _generate_time_slice!(m::Model)
     _generate_time_slice_relationships!(m::Model)
+    rep_time_slice_mapping(m::Model)
```
I think, for consistency this should be called `_generate_representative_time_slice_mapping()`.

Also, I'm not sure I'm a fan of abbreviating `representative` to `rep` in general?

manuelma commented on src/data_structure/temporal_structure.jl:
```diff
@@ -344,6 +345,28 @@ function to_time_slice(m::Model; t::TimeSlice)
+function rep_time_slice_mapping(m::Model)
+    rep_dict=Dict()
+    for blk in indices(ordering_periods)
+        for t_start_real in ordering_periods(temporal_block=blk).indexes
```
We'd need a check that `ordering_periods` is a TimeSeries to be able to use `indexes`. But also I'm not a fan of accessing fields directly...

manuelma commented:
I went through most of the code and I really like it. I suggested a bunch of renaming ideas and also asked a couple of minor questions that would be good to answer before merging, but apart from that, since test pass (as I believe they do) it's ok to merge.

We probably want to find time to add tests for the new functionality so we don't accidentally break in future pushes.

mihlema commented:
Thanks manuel for reviewing. I have addressed your suggestions and removed the "find conflicts functionality" and will submit another merge request on it. This should be good now.
